File: src/mutual/embed.py
"""Mutual — Embedding 生成。

对应 docs/engineering-plan.md §3.5、spec/02-stages.md §3。

生成 section + HyDE 向量，content-hash 驱动增量复用。
``embed_sections`` 是纯变换；``load_bundle`` / ``dump_bundle`` 是 adapter 用的磁盘 helper。

边界（spec/05-boundaries.md §6）：
- 复用是 **content-addressed**（``section_hashes``），不是 roster-addressed：
  改一个 profile 只重嵌该 profile 的变化 cell，不影响其他人。
- 不同 ``embedding_model`` 的 ``existing`` bundle 整体忽略（迁移守卫）；
  同名 model 但维度不一致时同样整体忽略（全量重嵌）。
- 全尺寸向量始终存储；MRL 截断在工作副本上做（``truncate_embeddings``）。
- 缓存/复用 key 用 ``hash_text``，禁止内置 ``hash()``。

spec 沉默：
- A-2：embed stage 的 input_schema（spec/02-stages.md §3）未声明 embedder
  注入方式。实现约定：可选 keyword 参数 ``embedder``（鸭子类型，
  ``embed(texts) -> [N, D]``，见 llm.EmbedderProtocol）；为 ``None`` 时尝试
  ``config["llm_wrapper"].get_embedding_model()``。
- A-3：``EmbeddingsBundle.user_timestamps`` 的来源在 embed 输入
  （``ExtractedSections`` 无时间戳字段）中不可得。实现：保留 ``existing``
  中已有用户的时间戳，新用户缺省不记录。
"""

from __future__ import annotations

import json
from typing import Any, Dict, List, Optional, Tuple

import numpy as np

from .extract import NOT_SPECIFIED
from .schemas import EmbeddingsBundle, HydeDescriptors, hash_text

_EPS = 1e-12


def embed_sections(
    sections: list,
    hyde: Dict[str, HydeDescriptors],
    config: Dict[str, Any],
    existing: Optional[EmbeddingsBundle] = None,
    embedder: Optional[Any] = None,
) -> EmbeddingsBundle:
    """生成 section + HyDE 向量；content-hash 驱动增量复用。

    所有可调参数从 ``config`` 读取（``models.embedding``、
    ``models.embedding_dimensions``），不硬编码。

    Args:
        sections: ``list[ExtractedSections]``。
        hyde: ``dict[user_id → HydeDescriptors]``。
        config: 配置 dict。
        existing: 已有 bundle，用于增量复用。``None`` 表示全量重嵌。
            若 ``existing.embedding_model`` 与 config 不一致，整体忽略（§6）。
        embedder: embedder 实例（鸭子类型 ``embed(texts) -> [N, D]``）。
            为 ``None`` 时从 ``config["llm_wrapper"]`` 获取（A-2）。

    Returns:
        :class:`~mutual.schemas.EmbeddingsBundle`，全尺寸存储；
        ``embeddings`` 形状 ``[N, sections, D]``，``hyde`` 形状
        ``{section: [N, n_desc, D]}``。

    边界：复用是 content-addressed 不是 roster-addressed（§6）。
    """
    model = config.get("models", {}).get("embedding", "")
    if embedder is None:
        wrapper = config.get("llm_wrapper")
        getter = getattr(wrapper, "get_embedding_model", None)
        if getter is not None:
            embedder = getter()
    if embedder is None:
        raise ValueError("embedder 未注入：传入 embedder 参数，或在 config 中提供 llm_wrapper")

    user_ids = [es.id for es in sections]
    section_names = sorted({name for es in sections for name in es.sections})
    section_index = {name: k for k, name in enumerate(section_names)}

    reuse = existing is not None and existing.embedding_model == model
    old_user_index: Dict[str, int] = {}
    old_section_index: Dict[str, int] = {}
    if reuse:
        assert existing is not None
        old_user_index = {uid: k for k, uid in enumerate(existing.user_ids)}
        old_section_index = {name: k for k, name in enumerate(existing.section_names)}

    def _plan(reuse_enabled: bool):
        """构建 cell 级嵌入计划。

        Returns:
            base_plan: ``{(user_i, section): ("reuse", old_i, old_si, hash)
            | ("new", text_pos, hash) | None}``；``None`` = 缺失 cell（零向量）。
            hyde_plan: ``{(user_i, section, k): ("reuse", old_i, old_k, hash)
            | ("new", text_pos, hash)}``。
            texts: 待嵌入文本（顺序对应 ``text_pos``）。
        """
        existing_ref = existing if reuse_enabled else None
        base_plan: Dict[Tuple[int, str], Any] = {}
        hyde_plan: Dict[Tuple[int, str, int], Any] = {}
        texts: List[str] = []
        for i, es in enumerate(sections):
            uid = es.id
            for name in section_names:
                content = es.sections.get(name, "")
                present = bool(content) and content != NOT_SPECIFIED
                if not present:
                    base_plan[(i, name)] = None
                    continue
                h = hash_text(content)
                reused = (
                    existing_ref is not None
                    and uid in old_user_index
                    and name in old_section_index
                    and existing_ref.section_hashes.get(f"{uid}|{name}") == h
                )
                if reused:
                    base_plan[(i, name)] = (
                        "reuse",
                        old_user_index[uid],
                        old_section_index[name],
                        h,
                    )
                else:
                    base_plan[(i, name)] = ("new", len(texts), h)
                    texts.append(content)
            hd = hyde.get(uid)
            if hd is None:
                continue
            for name, descs in hd.descriptors.items():
                if name not in section_index:
                    continue
                for k, desc in enumerate(descs):
                    h = hash_text(desc)
                    old_hyde = existing_ref.hyde.get(name) if existing_ref else None
                    reused = (
                        old_hyde is not None
                        and existing_ref is not None
                        and uid in old_user_index
                        and k < old_hyde.shape[1]
                        and existing_ref.hyde_hashes.get(f"{uid}|{name}|{k}") == h
                    )
                    if reused:
                        hyde_plan[(i, name, k)] = ("reuse", old_user_index[uid], k, h)
                    else:
                        hyde_plan[(i, name, k)] = ("new", len(texts), h)
                        texts.append(desc)
        return base_plan, hyde_plan, texts

    base_plan, hyde_plan, texts = _plan(reuse)
    vecs = embedder.embed(texts) if texts else None
    if vecs is not None:
        dim = int(vecs.shape[1])
    elif reuse and existing is not None:
        dim = int(existing.dim)
    else:
        dim = 0

    if reuse and existing is not None and vecs is not None and dim != int(existing.dim):
        # 迁移守卫：同名 model 但维度变化 → existing 整体忽略，全量重嵌。
        reuse = False
        base_plan, hyde_plan, texts = _plan(False)
        vecs = embedder.embed(texts) if texts else None
        dim = int(vecs.shape[1]) if vecs is not None else 0

    n_users = len(user_ids)
    n_sections = len(section_names)

    def _vec(arr: Any, pos: int) -> np.ndarray:
        return np.asarray(arr)[pos]

    embeddings = np.zeros((n_users, n_sections, dim), dtype=float)
    section_hashes: Dict[str, str] = {}
    for (i, name), cell in base_plan.items():
        if cell is None:
            continue
        if cell[0] == "reuse":
            assert existing is not None
            embeddings[i, section_index[name]] = existing.embeddings[cell[1], cell[2]]
            h = cell[3]
        else:
            embeddings[i, section_index[name]] = _vec(vecs, cell[1])
            h = cell[2]
        section_hashes[f"{user_ids[i]}|{name}"] = h

    n_desc: Dict[str, int] = {name: 0 for name in section_names}
    for hd in hyde.values():
        for name, descs in hd.descriptors.items():
            if name in n_desc:
                n_desc[name] = max(n_desc[name], len(descs))
    hyde_arrays: Dict[str, np.ndarray] = {
        name: np.zeros((n_users, count, dim), dtype=float) for name, count in n_desc.items()
    }
    hyde_hashes: Dict[str, str] = {}
    for (i, name, k), cell in hyde_plan.items():
        if cell[0] == "reuse":
            assert existing is not None
            hyde_arrays[name][i, k] = existing.hyde[name][cell[1], cell[2]]
            h = cell[3]
        else:
            hyde_arrays[name][i, k] = _vec(vecs, cell[1])
            h = cell[2]
        hyde_hashes[f"{user_ids[i]}|{name}|{k}"] = h

    user_timestamps: Dict[str, str] = (
        dict(existing.user_timestamps) if existing is not None and reuse else {}
    )

    return EmbeddingsBundle(
        user_ids=user_ids,
        section_names=section_names,
        embeddings=embeddings,
        hyde=hyde_arrays,
        embedding_model=model,
        dim=dim,
        section_hashes=section_hashes,
        hyde_hashes=hyde_hashes,
        user_timestamps=user_timestamps,
    )


def supports_mrl(model: str) -> bool:
    """检查 embedding model 是否支持 Matryoshka Representation Learning。

    支持 MRL 的模型（如 ``text-embedding-3-*``）可安全截断到更低维度；
    不支持的模型截断会破坏语义，应回退到全尺寸。

    spec 沉默 A-12：MRL 模型清单未在 spec/config 中定义。实现按 OpenAI
    ``text-embedding-3-*`` 家族前缀判定。

    Args:
        model: embedding model 标识（来自 ``config["models"]["embedding"]``）。

    Returns:
        是否支持 MRL 截断。
    """
    return model.startswith("text-embedding-3")


def truncate_embeddings(embeddings: np.ndarray, dim: int) -> np.ndarray:
    """MRL 截断：在全尺寸工作副本上截断到 ``dim`` 维。

    全尺寸向量始终存储；截断只发生在计算时的副本上（spec/05-boundaries.md §6）。
    截断后做 L2 归一化（零向量保持为零）。

    Args:
        embeddings: 全尺寸向量，形状 ``[..., D_full]``。
        dim: 目标维度（来自 ``config["models"]["embedding_dimensions"]``）。

    Returns:
        截断后的向量，形状 ``[..., dim]``。
    """
    if dim <= 0:
        raise ValueError(f"dim 必须为正整数，got {dim}")
    truncated = np.array(embeddings[..., :dim], dtype=float, copy=True)
    norms = np.linalg.norm(truncated, axis=-1, keepdims=True)
    return truncated / np.where(norms > _EPS, norms, 1.0)


def load_bundle(path: str) -> EmbeddingsBundle:
    """从磁盘加载 ``EmbeddingsBundle``（adapter 用）。

    Args:
        path: bundle 文件路径（npz + 元数据）。

    Returns:
        :class:`~mutual.schemas.EmbeddingsBundle`。
    """
    with np.load(path, allow_pickle=False) as z:
        meta = json.loads(str(z["meta"][()]))
        embeddings = np.asarray(z["embeddings"])
        hyde = {
            key.split("hyde::", 1)[1]: np.asarray(z[key])
            for key in z.files
            if key.startswith("hyde::")
        }
    return EmbeddingsBundle(
        user_ids=list(meta["user_ids"]),
        section_names=list(meta["section_names"]),
        embeddings=embeddings,
        hyde=hyde,
        embedding_model=meta["embedding_model"],
        dim=int(meta["dim"]),
        section_hashes=meta.get("section_hashes", {}),
        hyde_hashes=meta.get("hyde_hashes", {}),
        user_timestamps=meta.get("user_timestamps", {}),
    )


def dump_bundle(bundle: EmbeddingsBundle, path: str) -> None:
    """写入 ``EmbeddingsBundle`` 到磁盘（adapter 用，全尺寸存储）。

    Args:
        bundle: 待持久化的 bundle。
        path: 目标文件路径。
    """
    meta = json.dumps(
        {
            "user_ids": bundle.user_ids,
            "section_names": bundle.section_names,
            "embedding_model": bundle.embedding_model,
            "dim": int(bundle.dim),
            "section_hashes": bundle.section_hashes,
            "hyde_hashes": bundle.hyde_hashes,
            "user_timestamps": bundle.user_timestamps,
        },
        ensure_ascii=False,
    )
    arrays: Dict[str, Any] = {
        "embeddings": np.asarray(bundle.embeddings),
        "meta": np.array(meta),
    }
    for name, arr in bundle.hyde.items():
        arrays[f"hyde::{name}"] = np.asarray(arr)
    with open(path, "wb") as fh:
        np.savez(fh, **arrays)

File: internal/engine/embed.go
package engine

import (
	"fmt"
	"math"
	"sort"
	"strings"

	"github.com/Cloudbird-Software/mutual/internal/domain"
)

// Embedder 是 engine 对 embedding 服务的最小依赖（consumer-defined
// interface，与 LLMClient 同构）：texts 与返回向量一一对应。
type Embedder interface {
	// Embed 对一组文本生成向量 [N][D]。len(result) 必须等于
	// len(texts)，且各行维度一致——违反即实现 bug，调用方 fail loud。
	Embed(texts []string) [][]float64
}

// EmbedSections 生成 section + HyDE 向量（embed 阶段，纯变换；
// embedder 注入，无网络 IO）。
//
// 复用边界（spec/05-boundaries.md §6）：
//   - 复用是 content-addressed（SectionHashes 按 "user|section" 记录
//     内容 hash），不是 roster-addressed——改一个画像只重嵌该画像的
//     变化 cell，不影响其他人；
//   - existing.EmbeddingModel 与 model 不同 → existing 整体忽略
//     （迁移守卫）；同名 model 但维度变化 → 同样整体忽略，全量重嵌；
//   - 全尺寸向量始终存储；MRL 截断在计算副本上做（TruncateVectors）。
//
// 缺失 cell（NotSpecified / 空内容）为零向量，不产生嵌入调用。
func EmbedSections(
	sections []domain.ExtractedSections,
	hyde map[domain.UserID]domain.HydeDescriptors,
	model string,
	existing *domain.EmbeddingsBundle,
	embedder Embedder,
) (*domain.EmbeddingsBundle, error) {
	if embedder == nil {
		return nil, fmt.Errorf("embedder 未注入：EmbedSections 需要 Embedder 实现")
	}

	userIDs := make([]domain.UserID, len(sections))
	nameSet := map[string]bool{}
	for i, es := range sections {
		userIDs[i] = es.ID
		for name := range es.Sections {
			nameSet[string(name)] = true
		}
	}
	sectionNames := make([]string, 0, len(nameSet))
	for name := range nameSet {
		sectionNames = append(sectionNames, name)
	}
	sort.Strings(sectionNames)
	sectionIndex := make(map[string]int, len(sectionNames))
	for k, name := range sectionNames {
		sectionIndex[name] = k
	}

	reuse := existing != nil && existing.EmbeddingModel == model

	basePlan, hydePlan, texts := planCells(sections, hyde, existing, reuse, sectionIndex)
	vecs, dim, err := embedTexts(embedder, texts, existing, reuse)
	if err != nil {
		return nil, err
	}

	// 迁移守卫：同名 model 但维度变化 → existing 整体忽略，全量重嵌。
	if reuse && existing != nil && vecs != nil && dim != existing.Dim {
		reuse = false
		basePlan, hydePlan, texts = planCells(sections, hyde, nil, false, sectionIndex)
		vecs, dim, err = embedTexts(embedder, texts, existing, false)
		if err != nil {
			return nil, err
		}
	}

	nUsers := len(userIDs)
	nSections := len(sectionNames)

	embeddings := make(domain.EmbeddingTensor, nUsers)
	for i := range embeddings {
		embeddings[i] = make(domain.UserEmbeddings, nSections)
		for k := range embeddings[i] {
			embeddings[i][k] = domain.SectionEmbeddings{make(domain.Vector, dim)}
		}
	}

	sectionHashes := map[string]string{}
	for key, cell := range basePlan {
		if cell == nil {
			continue // 缺失 cell：零向量，不记 hash
		}
		var vec domain.Vector
		if cell.reuse {
			vec = copyVector(existing.Embeddings[cell.oldUser][cell.oldSection][0])
		} else {
			vec = copyVector(domain.Vector(vecs[cell.textPos]))
		}
		embeddings[key.user][sectionIndex[key.section]] = domain.SectionEmbeddings{vec}
		sectionHashes[string(userIDs[key.user])+"|"+key.section] = cell.hash
	}

	// 每个 section 的描述符槽位数 = 全体用户的最大值（不足补零向量）。
	nDesc := make(map[string]int, len(sectionNames))
	for _, name := range sectionNames {
		nDesc[name] = 0
	}
	for _, hd := range hyde {
		for name, descs := range hd.Descriptors {
			if _, known := nDesc[string(name)]; known && len(descs) > nDesc[string(name)] {
				nDesc[string(name)] = len(descs)
			}
		}
	}
	hydeArrays := make(map[domain.SectionName][][]domain.Vector, len(nDesc))
	hydeHashes := map[string]string{}
	for name, count := range nDesc {
		arr := make([][]domain.Vector, nUsers)
		for i := range arr {
			row := make([]domain.Vector, count)
			for k := range row {
				row[k] = make(domain.Vector, dim)
			}
			arr[i] = row
		}
		hydeArrays[domain.SectionName(name)] = arr
	}
	for key, cell := range hydePlan {
		var vec domain.Vector
		if cell.reuse {
			vec = copyVector(existing.Hyde[domain.SectionName(key.section)][cell.oldUser][cell.oldSection])
		} else {
			vec = copyVector(domain.Vector(vecs[cell.textPos]))
		}
		hydeArrays[domain.SectionName(key.section)][key.user][key.desc] = vec
		hydeHashes[string(userIDs[key.user])+"|"+key.section+"|"+itoa(key.desc)] = cell.hash
	}

	userTimestamps := map[string]string{}
	if reuse && existing != nil {
		for k, v := range existing.UserTimestamps {
			userTimestamps[k] = v
		}
	}

	return &domain.EmbeddingsBundle{
		UserIDs:        userIDs,
		SectionNames:   toSectionNames(sectionNames),
		Embeddings:     embeddings,
		Hyde:           hydeArrays,
		EmbeddingModel: model,
		Dim:            dim,
		SectionHashes:  sectionHashes,
		HydeHashes:     hydeHashes,
		UserTimestamps: userTimestamps,
	}, nil
}

// cellKey / hydeCellKey 是嵌入计划的最小定位单元。
type cellKey struct {
	user    int
	section string
}

type hydeCellKey struct {
	user    int
	section string
	desc    int
}

// embedCell 是一个待嵌入（或复用）的单元：
// reuse=true 时从 existing 的 (oldUser, oldSection) 拷贝；
// 否则取新嵌入结果的 textPos 行。hash 是该 cell 的内容 hash。
type embedCell struct {
	reuse      bool
	oldUser    int
	oldSection int
	textPos    int
	hash       string
}

// planCells 构建 cell 级嵌入计划。
//
// 返回 basePlan（分节 cell，值 nil = 缺失）、hydePlan（描述符 cell）、
// texts（待嵌入文本，顺序对应 textPos）。reuse=false 时 existing
// 视为不存在（整体忽略，迁移守卫路径）。
func planCells(
	sections []domain.ExtractedSections,
	hyde map[domain.UserID]domain.HydeDescriptors,
	existing *domain.EmbeddingsBundle,
	reuse bool,
	sectionIndex map[string]int,
) (map[cellKey]*embedCell, map[hydeCellKey]*embedCell, []string) {
	basePlan := map[cellKey]*embedCell{}
	hydePlan := map[hydeCellKey]*embedCell{}
	var texts []string

	oldUserIndex := map[domain.UserID]int{}
	oldSectionIndex := map[string]int{}
	if reuse && existing != nil {
		for k, uid := range existing.UserIDs {
			oldUserIndex[uid] = k
		}
		for k, name := range existing.SectionNames {
			oldSectionIndex[string(name)] = k
		}
	}

	for i, es := range sections {
		uid := es.ID
		for _, name := range sortedSectionKeys(sectionIndex) {
			content := es.Sections[domain.SectionName(name)]
			if content == "" || content == NotSpecified {
				basePlan[cellKey{i, name}] = nil
				continue
			}
			hash := domain.HashText(content)
			if reuse && existing != nil {
				if oldI, okU := oldUserIndex[uid]; okU {
					if oldK, okS := oldSectionIndex[name]; okS {
						if existing.SectionHashes[string(uid)+"|"+name] == hash {
							basePlan[cellKey{i, name}] = &embedCell{
								reuse: true, oldUser: oldI, oldSection: oldK, hash: hash,
							}
							continue
						}
					}
				}
			}
			basePlan[cellKey{i, name}] = &embedCell{textPos: len(texts), hash: hash}
			texts = append(texts, content)
		}
		hd, ok := hyde[uid]
		if !ok {
			continue
		}
		for _, name := range sortedDescriptorKeys(hd.Descriptors) {
			if _, known := sectionIndex[string(name)]; !known {
				continue
			}
			oldSlots := 0
			if reuse && existing != nil {
				if arr, has := existing.Hyde[name]; has && len(arr) > 0 {
					oldSlots = len(arr[0]) // 全体用户统一槽位数（构造保证）
				}
			}
			for k, desc := range hd.Descriptors[name] {
				hash := domain.HashText(desc)
				if reuse && existing != nil && k < oldSlots {
					if oldI, okU := oldUserIndex[uid]; okU {
						if existing.HydeHashes[string(uid)+"|"+string(name)+"|"+itoa(k)] == hash {
							hydePlan[hydeCellKey{i, string(name), k}] = &embedCell{
								reuse: true, oldUser: oldI, oldSection: k, hash: hash,
							}
							continue
						}
					}
				}
				hydePlan[hydeCellKey{i, string(name), k}] = &embedCell{textPos: len(texts), hash: hash}
				texts = append(texts, desc)
			}
		}
	}
	return basePlan, hydePlan, texts
}

// embedTexts 调 embedder 嵌入 texts；空 texts 返回 (nil, 继承维度)。
// 维度继承：texts 为空且复用时沿用 existing.Dim（无新向量可推断）。
func embedTexts(embedder Embedder, texts []string, existing *domain.EmbeddingsBundle, reuse bool) ([][]float64, int, error) {
	if len(texts) == 0 {
		if reuse && existing != nil {
			return nil, existing.Dim, nil
		}
		return nil, 0, nil
	}
	vecs := embedder.Embed(texts)
	if len(vecs) != len(texts) {
		return nil, 0, fmt.Errorf("embedder 契约违反：请求 %d 条文本，返回 %d 条向量", len(texts), len(vecs))
	}
	dim := 0
	if len(vecs) > 0 {
		dim = len(vecs[0])
		for _, row := range vecs {
			if len(row) != dim {
				return nil, 0, fmt.Errorf("embedder 契约违反：向量维度不一致（%d vs %d）", len(row), dim)
			}
		}
	}
	return vecs, dim, nil
}

func copyVector(v domain.Vector) domain.Vector {
	return append(domain.Vector(nil), v...)
}

func sortedSectionKeys(sectionIndex map[string]int) []string {
	names := make([]string, 0, len(sectionIndex))
	for name := range sectionIndex {
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}

func sortedDescriptorKeys(m map[domain.SectionName][]string) []domain.SectionName {
	names := make([]domain.SectionName, 0, len(m))
	for name := range m {
		names = append(names, name)
	}
	sort.Slice(names, func(i, j int) bool { return names[i] < names[j] })
	return names
}

func toSectionNames(names []string) []domain.SectionName {
	out := make([]domain.SectionName, len(names))
	for i, n := range names {
		out[i] = domain.SectionName(n)
	}
	return out
}

// SupportsMRL 判定 embedding model 是否支持 Matryoshka 截断
// （spec 沉默 A-12：按 OpenAI text-embedding-3-* 家族前缀判定）。
func SupportsMRL(model string) bool {
	return strings.HasPrefix(model, "text-embedding-3")
}

// TruncateVectors MRL 截断：全尺寸向量截断到 dim 维并 L2 归一化
// （零向量保持为零）。截断只发生在计算副本上（§6：存储恒全尺寸）。
func TruncateVectors(vectors []domain.Vector, dim int) ([]domain.Vector, error) {
	if dim <= 0 {
		return nil, fmt.Errorf("dim 必须为正整数，got %d", dim)
	}
	out := make([]domain.Vector, len(vectors))
	for i, v := range vectors {
		if dim > len(v) {
			out[i] = append(domain.Vector(nil), v...)
			continue
		}
		truncated := append(domain.Vector(nil), v[:dim]...)
		norm := 0.0
		for _, x := range truncated {
			norm += x * x
		}
		norm = math.Sqrt(norm)
		if norm > eps {
			for k := range truncated {
				truncated[k] /= norm
			}
		}
		out[i] = truncated
	}
	return out, nil
}
